src/onnx/parse_attention.cpp

```diff
@@ -0,0 +1,967 @@
+    static instruction_ref generate_raw_mask_per_batch(const onnx_parser::node_info& info,
+                                                       const instruction_ref& input,
+                                                       const instruction_ref& mask_input,
+                                                       const attention_attr& parsed_in,
+                                                       const attention_inferred& inferred_in)
+    {
+        auto batch_size    = inferred_in.batch_size;
+        auto total_seq_len = inferred_in.total_sequence_length;
+        auto num_heads     = parsed_in.num_heads;
+
+        // Other two cases require us to generate masks from sequence or total sequence length pads.
+        auto pass_value_lit = info.add_literal(
+            migraphx::literal{migraphx::shape{input->get_shape().type(), {1}, {1}}, {0}});
+        auto mask_value_lit = info.add_literal(migraphx::literal{
+            migraphx::shape{input->get_shape().type(), {1}, {1}}, {parsed_in.mask_filter_val}});
+
+        // For dim = 2 or dim =3 generate the apporiate mask across batches
+        // We need to handle the batch case since raw masking involes shape [batch, seq_len] or
+        // [batch, seq_len, total_seq_len],
+        auto bc_pass = info.add_instruction(
+            make_op("multibroadcast",
+                    {{"out_lens", {batch_size, num_heads, total_seq_len, total_seq_len}}}),
+            pass_value_lit);
+        auto bc_mask = info.add_instruction(
+            make_op("multibroadcast",
+                    {{"out_lens", {batch_size, num_heads, total_seq_len, total_seq_len}}}),
+            mask_value_lit);
+
+        auto raw_mask = mask_input;
+        // For raw masks we just need to mask out key value padding thus the 3d mask isn't needed
+        // here.
+        raw_mask = info.add_instruction(
+            make_op("reshape", {{"dims", {batch_size, 1, 1, total_seq_len}}}), raw_mask);
+        raw_mask = info.add_instruction(
+            make_op("multibroadcast",
+                    {{"out_lens", {batch_size, num_heads, total_seq_len, total_seq_len}}}),
+            raw_mask);
+        raw_mask = info.add_instruction(
+            make_op("reshape", {{"dims", {batch_size, num_heads, total_seq_len, total_seq_len}}}),
+            raw_mask);
+
+        // Reuse "0" broadcasted converted to int32 to check if input mask is greater than 0 for
+        // where condition
+        auto in_pass = info.add_instruction(
+            make_op("convert", {{"target_type", mask_input->get_shape().type()}}), bc_pass);
+        auto in_bool = info.add_instruction(make_op("equal"), raw_mask, in_pass);
+        /* in_bool = check_and_set_unidirectional(info, in_bool, parsed_in.unidirectional); */
+        // Need this to let MLIR to run with where
+        in_bool = info.add_instruction(
+            make_op("convert", {{"target_type", migraphx::shape::int8_type}}), in_bool);
```

Comment:
The convert should not be added here. If its needed just for mlir then this should be done in mlir.cpp.

Reply:
Sure gues that means I need another matcher then

where(x, y, z) and match when x.type() == bool_type to convert then.

Reply:
I dont know if a matcher is needed per se. You can just extend the `rewrite_reduce` function in mlir.cpp for the `where`. Probably need to rename the `rewrite_reduce` function to `rewrite_ops` as well.